src/main/java/com/vinylteam/vinyl/security/impl/DefaultSecurityService.java

```diff
@@ -0,0 +1,51 @@
+package com.vinylteam.vinyl.security.impl;
+
+import com.vinylteam.vinyl.entity.Role;
+import com.vinylteam.vinyl.entity.User;
+import com.vinylteam.vinyl.security.SecurityService;
+
+import javax.crypto.SecretKey;
+import javax.crypto.SecretKeyFactory;
+import javax.crypto.spec.PBEKeySpec;
+import java.security.NoSuchAlgorithmException;
+import java.security.SecureRandom;
+import java.security.spec.InvalidKeySpecException;
+import java.util.Arrays;
+import java.util.Random;
+
+public class DefaultSecurityService implements SecurityService {
+
+
+    @Override
+    public String hashPassword(char[] password, byte[] salt, int iterations) {
+
+        try {
+            SecretKeyFactory skf = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA512");
+            PBEKeySpec spec = new PBEKeySpec(password, salt, iterations, 256);
+            SecretKey key = skf.generateSecret(spec);
+            byte[] res = key.getEncoded();
+            return Arrays.toString(res);
+        } catch (NoSuchAlgorithmException | InvalidKeySpecException e) {
+            throw new RuntimeException("Exception while hashing password", e);
+        }
+    }
+
+    @Override
+    public byte[] generateSalt() {
+        Random r = new SecureRandom();
+        byte[] salt = new byte[20];
+        r.nextBytes(salt);
+        return salt;
+    }
```

Comment:
Please use whole words for variable naming

Reply:
Please don't create SecureRandom each time, if it's unnecessary

Reply:
> 
> 
> Please use whole words for variable naming

Okay

Reply:
> 
> 
> Please don't create SecureRandom each time, if it's unnecessary

Reasonable!

Reply:
> 
> 
> > Please don't create SecureRandom each time, if it's unnecessary
> 
> Reasonable!

@AlexKievDeveloper @TarasOrlovskyi Should I also get SecretKeyFactory out of the method?

Reply:
Yes I think that you can create SecretKeyFactory as private field in your class and use it in your method.

Reply:
Unfortunately, if I get try to initialise field secretKeyFactory outside method it causes "Unhandled exception: java.security.NoSuchAlgorithmException".

Should I instead make a try-with-resources and make secretKeyFactory a resource?

Reply:
> Unfortunately, if I get try to initialise field secretKeyFactory outside method it causes "Unhandled exception: java.security.NoSuchAlgorithmException".
> 
> Should I instead make a try-with-resources and make secretKeyFactory a resource?

Yes, something like it:

    private SecretKeyFactory skf;

    {
        try {
            skf = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA512");
        } catch (NoSuchAlgorithmException e) {
            throw new RuntimeException(".....", e);
        }
    }